### PR TITLE
Add cache, update several Emmet commands

### LINKED ISSUE #89821: Emmet match tag fails when tags contain comments jsx
Version: 1.41.1
Commit: 26076a4de974ead31f97692a0d32f90d735645c0
Date: 2019-12-18T15:04:31.999Z
Electron: 6.1.5
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Linux x64 5.0.0-37-generic snap

Command `"editor.emmet.action.matchTag"`

1. Can't jump to matching tag when the tag has a comment on an attribute. Matching still works with other tags such as the child div.
```js
import React from "react";

const Column = () => {
  return (
   {/* Can Not Jump */}
    <div
      className="parent"
      // onClick={() => {}}
      onAbort={() => {}}
      onAbortCapture={() => {}}
    > 
      <div className="child">title</div> {/* Can Jump */}
    </div>
  );
};

export default Column;
```
2. Inferred Types will disable any tag matching. This example is in a tsx file.
```js
import React, { useRef } from "react";

const Column = () => {
  const ref1 = useRef<HTMLElement>(null); // Inferred Type
  
  return (
   {/* Can Not Jump */}
    <div
      className="parent"
      onClick={() => {}}
      onAbort={() => {}}
      onAbortCapture={() => {}}
    > 
      <div className="child">title</div>{/* Can Not Jump */}
    </div>
  );
};

export default Column;

```

### LINKED ISSUE #99896: Emmet balance, updatetag failing due to invalid HTML parsing algorithm
- VSCode Version: 1.46.0
- OS Version: 10.15.5

Steps to Reproduce:

1. Create a `tsx` file with this content:
   ```tsx
   import React from 'react';

   // <f

   const MyComponent = () => <div className="foo">Hello, World</div>;
   ```
2. Highlight the opening tag of `div` and run any one of these commands:
   - "Emmet: Split/Join Tag"
   - "Emmet: Update Tag"
   - "Emmet: Go to Matching Pair"

Expected: the command should work
Actual: the command does nothing

If we remove the comment that contains `<f`, it works.

Does this issue occur when all extensions are disabled?: Yes

### LINKED ISSUE #112150: Emmet balance commands not working on html files with unclosed tag
Issue Type: <b>Bug</b>

1. Open a file like:
```
<html>
  <body>
    <p>
      First part of file
    </p>
```
2. Place cursor at opening `<p>`
2. Open command palette
3. Run Emmet command; eg 'Emmet: Balance (Outward)'

This should highlight from the `<p>` to `</p>` inclusive. Currently does nothing with no error either appearing or in Developer Console.
After closing `<body>` and `<html>` functions as expected.

This occurs with all extensions disabled, and in the latest Insiders build.

VS Code version: Code 1.51.1 (e5a624b788d92b8d34d1392e4c4d9789406efe8f, 2020-11-10T23:34:32.027Z)
OS version: Windows_NT x64 10.0.19042

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-6800K CPU @ 3.40GHz (12 x 3400)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>opengl: enabled_on<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.90GB (19.37GB free)|
|Process Argv|--disable-extensions|
|Screen Reader|no|
|VM|0%|
</details>Extensions disabled

### PR DESCRIPTION
**Pull request by rzhao271**
This PR fixes #99896 and affects #71996 and #110577

### CODE REVIEW

**Comment by rzhao271** [comment]
This PR should also fix #89821

File: extensions/emmet/src/matchTag.ts [hunk 1]
```diff
@@ -4,42 +4,46 @@
...
+	const currentNode = getHtmlNodeLS(document, position, true);
+	if (!currentNode) {
+		return;
+	}
+
+	const offset = document.offsetAt(position);
 
 	// If no closing tag or cursor is between open and close tag, then no-op
-	if (!currentNode.close || (position.isAfter(currentNode.open.end) && position.isBefore(currentNode.close.start))) {
+	if (!currentNode?.endTagStart
```
Comment (aeschli) [inline extensions/emmet/src/matchTag.ts:40]: currentNode is defined here, so the '?' can be removed.
I thory 0 is a valid offset, admittedly not possible for the tagStart. So better `currentNode.endTagStart === undefined`

File: extensions/emmet/src/emmetCommon.ts [hunk 2]
```diff
@@ -145,6 +146,14 @@ export function activateEmmetExtension(context: vscode.ExtensionContext) {
 			updateEmmetExtensionsPath(true);
 		}
 	}));
+
+	context.subscriptions.push(vscode.workspace.onDidOpenTextDocument((e) => {
+		addFileToParseCache(toLSTextDocument(e));
```
Comment (aeschli) [inline extensions/emmet/src/emmetCommon.ts:151]: I found `addFileToParseCache` confusing. Looking at the implementation, it set and entry with undefined to the cache. Is that necessary? it's done for all URI, regardless if html or not.

**Comment by aeschli** [comment]
I glanced over it and it looks good!

**Review by aeschli — APPROVED** [review]

File: extensions/emmet/src/emmetCommon.ts (hunk 2 above)
Comment (rzhao271) [reply extensions/emmet/src/emmetCommon.ts:151]: Right, I should make it only save markup files like HTML or XML

**Pull request merged**